server/src/main/scala/com/cloudera/livy/server/recovery/FileSystemStateStore.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to Cloudera, Inc. under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Cloudera, Inc. licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.server.recovery
+
+import java.io.{ByteArrayOutputStream, FileNotFoundException}
+import java.net.URI
+import java.util
+
+import org.apache.hadoop.fs.{CreateFlag, FileContext, FSDataInputStream, Path}
+import org.apache.hadoop.fs.Options.{CreateOpts, Rename}
+
+import com.cloudera.livy.{LivyConf, Logging}
+
+object FileSystemStateStore extends StateStoreCompanion {
+  override def create(livyConf: LivyConf): StateStore = new FileSystemStateStore(livyConf)
+}
+
+class FileSystemStateStore(livyConf: LivyConf) extends StateStore with Logging {
+  private val fsUri = {
+    val fsPath = livyConf.get(LivyConf.RECOVERY_STATE_STORE_URL_CONF)
+    require(!fsPath.isEmpty, s"Please config ${LivyConf.RECOVERY_STATE_STORE_URL_CONF.key}.")
+    new URI(fsPath)
+  }
+
+  private val fileContext: FileContext = FileContext.getFileContext(fsUri)
+
+  override def set(key: String, value: Object): Unit = {
+    // Write to a temp file then rename to avoid file corruption if livy-server crashes
+    // in the middle of the write.
+    val tmpPath = absPath(s"$key.tmp")
+    val tmpFileOutput = fileContext.create(tmpPath,
+      util.EnumSet.of(CreateFlag.CREATE, CreateFlag.OVERWRITE),
+      CreateOpts.createParent())
+
+    try {
+      tmpFileOutput.write(serializeToBytes(value))
+    } finally {
+      tmpFileOutput.close()
+      // Assume rename is atomic.
+      fileContext.rename(tmpPath, absPath(key), Rename.OVERWRITE)
+    }
+  }
+
+  override def get[T](key: String, valueType: Class[T]): Option[T] = {
```

Comment:
But if your class instance is not multi-thread shared, I think it should be safe to use ClassTag. Also we only use ClassTag to store class information, don't use reflection API to get object, perhaps it is safe to use here.